Browse latest changes on the UI

dkocich opened the pull request:
Related to UI issue https://github.com/c2corg/v6_ui/issues/1323

Requires to merge/use API PR to work correctly https://github.com/c2corg/v6_api/pull/620

There is a new page `localhost:6558/whatsnew` which allows browsing of latest changes. ~~It uses [ngTable](http://ng-table.com/) and data are loaded in one request to API. It is possible to use filters, sort values, change page and its size.~~

![image](https://cloud.githubusercontent.com/assets/6165660/22547256/25dba5a0-e941-11e6-9b4e-e9217c6ae0d5.png)

tsauerwein commented on c2corg_ui/views/index.py:
```diff
@@ -57,6 +58,28 @@ def following(self):
             'following', 'c2corg_ui:templates/following.html')
 
+    @view_config(route_name='whatsnewdata', renderer='json')
+    def whatsnewdata(self):
```
Do you need this view? The request could be made directly to the API (instead of making a request to the UI server-side which makes a request to the API).

tsauerwein commented on c2corg_ui/static/js/whatsnew.js:
```diff
@@ -0,0 +1,94 @@
+   * @export
+   */
+  this.columns = [
+    {field: 'version_id', title: 'version_id', show: true},
```
To avoid that the Closure Compiler might rename the property names, you can either provide a definition in the externs file or use `'file': 'version_id'`.

tsauerwein commented on c2corg_ui/static/js/whatsnew.js:
```diff
@@ -0,0 +1,94 @@
+   * @export
+   */
+
+  this.dataset = [{}];
```
`this.dataset = [];`?

asaunier commented:
Does the API service really support cols sorting and pagination?
I see nothing about that in https://github.com/c2corg/v6_api/pull/620/files

dkocich commented:
> Does the API service really support cols sorting and pagination?

The higher amount of documents is loaded from API now and ordering/pagination/filtering is done with them on UI only. It can be useful to quickly search in the data without creating of additional requests to API. I started to work on pagination on API and I am not sure if there should be some/all filters implemented also. This is just first wip proposal so feel free to comment.

asaunier commented:
> The higher amount of documents is loaded from API now and ordering/pagination/filtering is done with them on UI only.

Is it really possible? There are about 850'000 documents in the database, probably millions of changes overall. I don't understand how we could handle this in the UI only?

asaunier commented:
By the way, are you sure sorting was really requested by the client? I rather think the goal is to be able to see the latest changes, ie. starting from the most recent down to older ones. Only a chronological sorting is needed.

Concerning the pagination, I would recommend to use something similar to the feed, with a direct infinite-scrolling request to the API (the UI is only used to display the page with the calling directive).

And if we could avoid relying on yet another library, it would save bytes to download :P

desnoes commented:
> By the way, are you sure sorting was really requested by the client? I rather think the goal is to be able to see the latest changes, ie. starting from the most recent down to older ones. Only a chronological sorting is needed.

I confirm, a chronological sorting is sufficient.

dkocich commented:
> Is it really possible? There are about 850'000 documents in the database, probably millions of changes overall. I don't understand how we could handle this in the UI only?

I tried to request 500-2500 documents and display them using ngTable. They were divided into pages based on the page size setting (bottom right in the previous picture). All the sorting, filtering, etc. worked only over one requested batch of documents and not over the complete table of course.

> @asaunier Concerning the pagination, I would recommend to use something similar to the feed, with a direct infinite-scrolling request to the API (the UI is only used to display the page with the calling directive).

I updated this PR. Changes are now listed as a feed into the table but JS to handle user filter is not yet implemented. If you would really like this approach more, then I can maybe try to add that user filter and improve some details to get this PR merged - what do you think? Thanks.

![image](https://cloud.githubusercontent.com/assets/6165660/22692506/b4abf7d8-ed3f-11e6-8739-538ef26a2062.png)

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,42 @@
+<section id="whatsnew">
+
+  <header>
+    <h1 class="text-center" translate><span>Feed of the latest changes</span></h1>
```
You don't need the span element here. Or at least the translate directive should apply to it, not to the parent, else the span tags are integrated to the string to translate.

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,42 @@
+  </div>
+
+  <div>
+    <app-feed app-is-changes-feed="true" template-path="/static/partials/feedtable.html"></app-feed>
```
Are you sure of this ``template-path`` attribute here? Shouldn't it be passed as ``templateUrl`` in the directive.
I'am afraid that using this attribute here will not make the partial available in the ``templatecache.js`` file.

dkocich replied:
I tried `make clean`, `make build` and I see `$templateCache.put('/static/partials/feedtable.html'...`,  in my `static/build/templatecache.js`

asaunier replied:
OK, perhaps that's because this file is built using all files in ``/static/partials``. But it doesn't mean this template cache will actually be used. It seems ``/static/partials/feedtable.html`` will always be called, including in prod mode.

asaunier replied:
OK I understand: you try to reuse the feed directive used for the home/profile feeds and would like to customize the partial to use...

I am not sure this is the right way. It makes things more complicated and finally all those feeds are different. It would make sense to use dedicated directives I think?!

To avoid duplicating the common code. You should create a parent controller class and use child classes for the specific parts. As we did for instance for the document editing.

asaunier commented on c2corg_ui/views/index.py:
```diff
@@ -57,6 +57,18 @@ def following(self):
+    @view_config(route_name='whatsnew',
+                 renderer='c2corg_ui:templates/whatsnew.html')
+    def whatsnew(self):
+        resp, data = call_api(self.settings, 'document/changes')
```
I don't understand why you call the API from within a UI server view instead of making a direct request from the browser, as done for the home/profile feeds.

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,42 @@
+  <div class="tools">
+
+##     <h3 class="heading show-phone" ng-click="mainCtrl.animateHeaderIcon($event)" data-toggle="collapse" data-target="#filters">
+##       <span translate>Filters</span><span class="glyphicon glyphicon-menu-down"></span>
```
I don't think we need to have filters.
A link in the user's profiles towards ``whatsnew?u=12345`` should be enough.

Perhaps the client will need advanced filtering on changes in the future. But it's not required for now.

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,42 @@
+  ## Set in the data request
+  module.value('mapFeatureCollection', null);
+
+</%block>
```
That's already the default value:
https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/templates/base.html#L153-L155
You don't need to override it.

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,69 @@
+        This feed contains {{amount}} changes.</p>
+
+      <tr>
+        <th>type</th>
```
translate

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,69 @@
+  <div class="table-responsive">
+    <table class="feedtable table table-condensed table-bordered table-striped table-hover">
+
+      <p class="text-success text-center" ng-if="feedCtrl.userId" ng-init="user = feedCtrl.userId;" ng-model="amount = feedCtrl.feedLength;">
```
Why ``ng-model="amount = feedCtrl.feedLength;"``?
It should be part of the ``ng-init`` I think.

dkocich replied:
It should not. `feedLength` is undefined on init and it contains a correct value after the response is received.

asaunier commented on c2corg_ui/static/js/feed.js:
```diff
@@ -204,4 +235,31 @@ app.FeedController.prototype.documentType = function(type) {
+ * @param doc {Object}
+ * @returns {string | boolean}
+ * @suppress {checkTypes}
+ * @constructor
```
Constructor?

asaunier commented on c2corg_ui/static/js/feed.js:
```diff
@@ -204,4 +235,31 @@ app.FeedController.prototype.documentType = function(type) {
+/**
+ * Generate URL address to the changed document version
+ * @param doc {Object}
+ * @returns {string | boolean}
```
It doesn't make sense that a function returns that unconsistent outputs. Make a choice :P

asaunier commented on c2corg_ui/static/js/feed.js:
```diff
@@ -204,4 +235,31 @@ app.FeedController.prototype.documentType = function(type) {
+ * Generate URL address to the changed document version
+ * @param doc {Object}
+ * @returns {string | boolean}
+ * @suppress {checkTypes}
```
??
Stashing problems under the carpet? :P

asaunier commented on c2corg_ui/static/js/feed.js:
```diff
@@ -204,4 +235,31 @@ app.FeedController.prototype.documentType = function(type) {
+
+/**
+ * Generate URL address to the changed document version
+ * @param doc {Object}
```
Do you mean ``appx.Document``?
https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/externs/appx.js#L111

asaunier commented on c2corg_ui/static/js/apiservice.js:
```diff
@@ -399,17 +399,23 @@ app.Api.prototype.updatePreferredLanguage = function(lang) {
  * @return {!angular.$q.Promise<!angular.$http.Response>}
  */
-app.Api.prototype.readFeed = function(token, lang, userId, isPersonal) {
+app.Api.prototype.readFeed = function(token, lang, userId, isPersonal, isChangesFeed) {
```
Since those are really different feeds, I would recommend to use dedicated functions.

asaunier commented:
I have set the [WIP] prefix back because it seems there is still a lot of work needed in this PR :P

asaunier commented:
I have run your branches in my instance. The tool works pretty well :)

![capture du 2017-02-10 15-13-48](https://cloud.githubusercontent.com/assets/1192331/22829610/adec5d0a-efa3-11e6-9ec2-1793136aab1b.png)

I see a few things that are not optimal and could be improved:
* ``document_id`` and ``version_id`` are not needed and could be removed
* make the title (or the whole line to make it easier to use on a mobile phone) clickable, pointing to the document archived version.
 * Perhaps pointing to the diff compared to the previous change would be even more relevant to figure the change out. But I guess we don't have the id of the previous version (perhaps a possible improvement in the future :P ).
* title should appear at the beginning of the line
* lang could be a suffix of the title, between parenthesis (as in cards)
* document types and qualities could be displayed using the icons as in cards.
* do we really need the columns titles? The content is quite straightforward, especially if we remove some cols and group some data
* the page is not usable on a mobile phone => the overflowing parts of the table are no longer visible => tables are not responsive :P 

Questions:
* where will this new "whatsnew" page be linked to? @stef74 @desnoes ?

brunobesson commented:
>tables are not responsive

There are several possibilities:
* bootstrap does it so that the tables can be scrolled horizontally (https://getbootstrap.com/css/#tables-responsive)
* change the layout like it has been done for snow conditions
* datatables hide some columns that can be expanded on purpose (https://datatables.net/extensions/responsive/), but it's a more complicated solution

asaunier commented:
> There are several possibilities:

Another possibilty: do we really need a table? That could be a line that would wrap if too long?

dkocich commented:
> Perhaps pointing to the diff compared to the previous change

I was also thinking about adding a link to the Diff between old and new docs... It would require changes to the API part to contain version's previous IDs...

> bootstrap does it so that the tables can be scrolled horizontally https://getbootstrap.com/css/#tables-responsive

[I did use it in this PR ](https://github.com/c2corg/v6_ui/pull/1466/files#diff-cf915b7eac979d3d23c4d6d4877eea28R11) but it probably don't work how it should :) ...

asaunier commented:
When the page first loads, the loading message is displayed twice. I guess only the message at the bottom of the feed is needed?!

@dkocich Thanks for the layout changes you have made. I find it pretty good now!
I would recommend to align the titles to the left, it's easier to read and parse than when they are centered.

asaunier commented on c2corg_ui/static/js/apiservice.js:
```diff
@@ -428,6 +428,28 @@ app.Api.prototype.readFeed = function(token, lang, userId, isPersonal) {
+  if (token) params['token'] = token;
+  if (userId) {
+    params['u'] = userId;
+  }
```
The coding standard of those last 3 lines is not very consistent.

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+          </span>
+        </td>
+
+        <td>{{::type}}</td>
```
Just an idea: what about using the document type badge classes (with the color) as in the document page titles?
https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/templates/helpers/view.html#L824

It would add some color to the list and make it easier to identify the doc type at first sight.

dkocich replied:
That badges look good. I added some colours to the types which had none selected.
![image](https://cloud.githubusercontent.com/assets/6165660/23059462/ca50a57c-f4f9-11e6-8dab-6ec7a5532656.png)

asaunier replied:
Nice! Let me know when the PR is ready for review.

asaunier replied:
> I added some colours to the types which had none selected.

Have you change the badge colors in the actual document pages too?
That would make sense, to have the same colors for badge in the change feed in the document detail views.

asaunier commented on c2corg_ui/static/js/whatsnewfeed.js:
```diff
@@ -0,0 +1,186 @@
+/**
+ * Generate URL address of a document
+ * @param doc {appx.DocumentChange}
+ * @returns {string}
```
Missing ``@export``

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+          ng-init="type = wfeedCtrl.documentType(doc.document.type); console.log(type);" >
+
+        <td class="text-left">
+          <a href="{{wfeedCtrl.getDocHref(doc)}}">{{::doc['document']['title']}}</a> ({{::doc['lang']}})
```
``ng-href``?
As for now the links are not working.

asaunier replied:
I would show the lang only if different from the current interface language.
A bit like https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/static/partials/cards/waypoints.html#L10

asaunier replied:
I see a problem with image with no title => it's not possible to click on the title because there is... no title :P 
We need a default (i18n) string when no title is available. Perhaps "Untitled image"?

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+    <table class="text-center feedtable table table-condensed table-striped table-hover">
+
+      <p class="text-success text-center" ng-if="wfeedCtrl.userId"
+         ng-init="user = wfeedCtrl.userId;" ng-model="amount = wfeedCtrl.feedLength;">
```
Missing translate directive

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+        This feed for the user {{user}} contains {{amount}} changes.
+      </p>
+      <p class="text-success text-center" ng-if="!wfeedCtrl.userId"
+         ng-model="amount = wfeedCtrl.feedLength;">
```
Missing translate directive

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+        <td class="text-left">
+          <a href="{{wfeedCtrl.getDocHref(doc)}}">{{::doc['document']['title']}}</a> ({{::doc['lang']}})
+          <span ng-if="doc['document']['version'] > 1 && (doc['document']['type'] !== 'i' && doc['document']['type'] !== 'o')">
+            (<a ng-href="{{wfeedCtrl.getVersionHref(doc)}}">previous version</a> | diff)
```
Does the "previous version" link really work? The link seems not clickable in my instance and the cursor remains a text cursor, not a pointer.

Actually the link toward the archived version should be the main one (when clicking the title).

For instance if the feed advertises 5 changes of the same document in a row, it does not make sense that the main links redirect to the current version of the document, they should point to the actual archived version.
Perhaps we don't even need to make a link toward the last version of the document!?

What about the "diff"? Do you think it is fairly feasible? If not, please remove it.

dkocich replied:
Prev. version works.

I think that it is better to have those links (to document + to the previous version) separately. In a case of documents which have just one version, it is better to point to that document otherwise an user has no chance to find it.
It can be confusing to sometimes have a title's `href` pointing to the previous version and sometimes for latest document (if it was just created)...

Now when there are those document badges I have an idea if it is better to have link to the document on the badge and if the document has any previous version -> point to it with the title (and delete "previous version" text). What do you think?

![image](https://cloud.githubusercontent.com/assets/6165660/23060725/4669ff46-f4ff-11e6-92a0-e0a44c7c51c3.png)

asaunier replied:
> I think that it is better to have those links (to document + to the previous version)

I am not sure to understand what you mean. Actually why do we need links towards several versions?

What do you call "previous version"? It seems your "previous version" links point to the archived version concerned by the change, doesn't it? This is actually the main link we want to see.

If you have several changes in a row of the same document, it's no use that the main links (on titles) point to the same URL...

The goal is to report that user X has created/edited a document at time T and to show what changes they have made. It should work a bit like the history pages (no need to compare several versions though) => point to the concerned version of the document and (if possible - not sure the previous version id is provided by the API response, is it?) to the diff of the changes that have been made (no diff anyway if this is a creation).
No need to point towards the last version of the document (or as a secondary link, could be named "current version").
No need to point towards the previous version of the document.

>  if it is better to have link to the document on the badge and if the document has any previous version

I don't know. In detail view pages, the badge points to the advanced search page. It would not be very consistent to make it point here to the document itself.
On the other hand, the badge name uses the singular form, not the plural form, so it's not really the same badge...

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+        <th class="text-center" translate>type</th>
+        <th class="text-center" translate>written_at</th>
+        <th class="text-center" translate>quality</th>
+        <th class="text-center" translate>username</th>
```
username => author

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+
+        <td class="text-left">
+          <a href="{{wfeedCtrl.getDocHref(doc)}}">{{::doc['document']['title']}}</a> ({{::doc['lang']}})
+          <span ng-if="doc['document']['version'] > 1 && (doc['document']['type'] !== 'i' && doc['document']['type'] !== 'o')">
```
Why do you have such a restriction on outings and images?

dkocich replied:
I wasn't sure how we store old versions of these documents because many created URL addresses of these document types weren't accessible. I simply hide them but it is possible that it was caused by some other problem. For example 1st version of this image can't be opened (probably created as private and later changed to collaborative?) https://www.camptocamp.org/images/history/851612/fr

asaunier replied:
> For example 1st version of this image can't be opened

That's weird. Ditto for any archive version of https://www.camptocamp.org/images/history/208579/fr
There is probably an error somewhere. I have created an issue at https://github.com/c2corg/v6_ui/issues/1503

I would recommend to drop those special cases: image archives should be accessible and outings will probably no longer be listed in the feed. https://github.com/c2corg/v6_api/issues/625

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+        </td>
+
+        <td data-original-title="999" data-container="body" tooltip-append-to-body="true" tooltip-placement="top"
+            uib-tooltip="user_id: {{::doc['user']['user_id']}} username: {{::doc['user']['username']}} ">
```
We don't need this (user id and username). All need info are provided with ``doc['user']['name']``.

asaunier commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -0,0 +1,66 @@
+            uib-tooltip="{{'quality'| translate}} : {{::doc['document']['quality']}}">
+        </td>
+
+        <td data-original-title="999" data-container="body" tooltip-append-to-body="true" tooltip-placement="top"
```
999?

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,18 @@
+<section id="whatsnew">
+
+  <header>
+    <h1 class="text-center" translate>Feed of the latest changes</h1>
```
What about simply "Latest changes"?

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,18 @@
+<%namespace file="helpers/common.html" import="show_title"/>
+
+<%block name="metarobots">
+  <meta name="robots" content="index,follow">
```
I don't think this page should be indexed. But no big deal.
Anyway this is already the default value: https://github.com/c2corg/v6_ui/blob/master/c2corg_ui/templates/base.html#L29
=> you can omit this block if you keep it as is.

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,18 @@
+<%inherit file="base.html"/>
+<%namespace file="helpers/common.html" import="show_title"/>
```
Where is ``show_title`` used?

asaunier replied:
On the other hand the HTML page title is not correct (as for now is is "home page") => please update the page title.

asaunier commented on less/feed.less:
```diff
@@ -193,3 +193,71 @@
+      }
+
+    }
+
```
Useless blank lines

asaunier commented on c2corg_ui/templates/whatsnew.html:
```diff
@@ -0,0 +1,18 @@
+  <meta name="robots" content="index,follow">
+</%block>
+
+<section id="whatsnew">
```
What about using a class instead?

asaunier commented on less/feed.less:
```diff
@@ -193,3 +193,71 @@
+  .feed-list {
+    transition: .3s;
+
+    .timeline-bullet {
```
Do you really use those styles? There is no timeline in the page, only a table.

dkocich commented:
I have no issue but travis fails :( https://travis-ci.org/c2corg/v6_ui/builds/202559775#L692-L708

```npm install
npm ERR! Linux 4.4.0-51-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v6.9.5/bin/node" "/home/travis/.nvm/versions/node/v6.9.5/bin/npm" "install"
npm ERR! node v6.9.5
npm ERR! npm  v3.10.10
npm ERR! shasum check failed for /tmp/npm-4869-28477fd8/registry.npmjs.org/moment/-/moment-2.15.2.tgz
npm ERR! Expected: 1bfdedf6a6e345f322fe956d5df5bd08a8ce84dc
npm ERR! Actual:   41b4e7ef38051e320e9b7b8f6eae35003f57e44c
npm ERR! From:     https://registry.npmjs.org/moment/-/moment-2.15.2.tgz
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm ERR! Please include the following file with any support request:
npm ERR!     /home/travis/build/c2corg/v6_ui/npm-debug.log```

asaunier commented:
I have just tested the last version of your branch, I see problems anyway:

* [ ] Your links towards the document detail views are not correct, their slug is defective.
For instance "http://c2corgv6-demo.gis.internal:6555/outings/842111/fr/Mont%20de%20Grange%20:%20Couloir%20de%20Pertuis"
* [ ] Title of routes are not correct: the ``title_prefix`` is missing.
* [ ] The mouse cursor on badges become a pointer (as for links) where there is no link.

As I said in my previous comment, I don't think using the standard document URL is relevant as the main link.

asaunier commented:
I have just tested the last version of the branch.

Why don't you put the link on the document title instead of using a "current version" link?
It makes more sense (and it's more handy) to put the link on the title.
In addition "current version" is a confusing term. For me it means "the most recent version", not the version of the link. 
Finally I see many entries with no link at all (not "current version" link)!? Is it because my API is not up-to-date?

What about putting the link on the whole line as I suggested it earlier?

What I recommend:
* drop "current version"
* make sure a title is provided (at least a default "Untitled" => should be translated then)
* badges should be translated
* it seems quality tooltips are not translated => "quality" is translated but not the attribute value
* lang data => only if different from the interface language (as in cards)

asaunier commented:
@dkocich With commit f1fe8ee I think I have fixed/simplified the links and main titles.

asaunier commented:
I see other minor problems I think. Perhaps I will suggest another commit.

dkocich commented on c2corg_ui/static/partials/feedtable.html:
```diff
@@ -24,12 +24,11 @@
-          <span ng-if="doc['lang'] != locale.lang">({{::doc['lang']}})</span>
+          <a ng-href="{{wfeedCtrl.getVersionUrl(doc)}}">
+            <span ng-if="doc['document']['title']">{{::doc['document']['title']}}</span>
+            <span ng-if="!doc['document']['title']" translate>Untitled</span>
```
I checked and there is again that problem with not-accessible image versions.

asaunier replied:
> I checked and there is again that problem with not-accessible image versions.

That's not our business here. It's an API issue: https://github.com/c2corg/v6_ui/issues/1503
Hopefully it will be fixed and the links will work here as well. But it does not make sense to make workarounds here, it's better to fix the problem at its source.

asaunier commented:
Argh, there is a problem with your changes (I have tried to remove mine with no more success): the home page feed is broken, only the initial list of items is loaded but when scrolling the feed is not refreshed.

asaunier commented on c2corg_ui/static/js/feed.js:
```diff
@@ -126,6 +126,7 @@ app.FeedController = function(appAuthentication, appApi, appLang, imageUrl, ngeo
  * Fills the feed with documents.
  * Used by ng-infinite-scroll directive in the template.
+ * @export
  * @public
```
Are you sure the ``@public`` keyword are needed?

dkocich replied:
it works without it just with `export`

asaunier commented:
Please do no more changes to this branch I was about to commit small problems fixes...

asaunier commented:
With commit 741501d I have made small improvements.

For me the PR is OK now, except that commits should be squashed (including mines, no problem).

dkocich commented:
@asaunier This pr is prepared to be merged.

desnoes commented:
@dkocich : nice job ! 
I have a question: what are the types of comments that should be displayed in the table ? I only observe 'creation'. Is it the expected behavior ?
I found a very minor bug: there is a 1 hour delay between the timestamp written in the table and the timestamp written in the tooltip. I would suggest to remove the tooltip on this field.

![bug whats new](https://cloud.githubusercontent.com/assets/19892089/23182689/bffb5666-f879-11e6-88cd-c331bdee3a33.png)

asaunier commented:
> I found a very minor bug: there is a 1 hour delay between the timestamp written in the table and the timestamp written in the tooltip.

I think the tooltip shows the GMT time.

> I would suggest to remove the tooltip on this field.

+1

desnoes commented:
I have removed the geometry in this outing and the modification is not reported.
http://www.demov6.camptocamp.org/outings/828402/en/tete-de-vauclave-rrr

dkocich commented:
> I have a question: what are the types of comments that should be displayed in the table? I only observe 'creation'. Is it the expected behaviour?

What would you like to see? When someone creates the document then it is "creation" automatically. When someone edits document then he is asked to write the short comment about the new version and to change quality.
@asaunier maybe we should add `translate` directive which would replace this word with something like "This document was just created." in all languages? It should be changed on doc. history pages too... 

![image](https://cloud.githubusercontent.com/assets/6165660/23183616/f13e9410-f87c-11e6-9989-ebda5c06671d.png)

>I would suggest to remove the tooltip on this field.

Ok.

> I have removed the geometry in this outing and the modification is not reported. http://www.demov6.camptocamp.org/outings/828402/en/tete-de-vauclave-rrr

I think that it is not possible to see any outing change - in this PR we removed User profile creations and Outing changes from the feed https://github.com/c2corg/v6_api/issues/625#issuecomment-280233276

desnoes commented:
I understood my mistake. The comment column report the text written in the field 'Comment about the changes'. Sorry for the inconvenience.

I consider the result is very fine. I close the issue and let you remove the tooltip on the timestamp.

dkocich commented:
No problem. :) Thanks!

fbunoz commented:
With this PR, is it possible to filter on the doc type, the lang, the linked areas, the activities (routes and outings only) ?

dkocich commented:
@fbunoz it is not possible right now
